### PR TITLE
pre defined classes-mathematical program-hard-q4

### PR DESCRIPTION
**Pull request by SamamthaH**
Wrote out clause for Calc Bill, awaiting double literals to finish giving clearer error messages to student

### CODE REVIEW

File: src/pre_defined_classes/mathematical_program/hard/q4/CalcBillTest.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        };
+    }
+
+    public void runMain() {
+        CalcBill.main(new String[]{});
+    }
+
+    @Test
+    public void tipPercentageTest() {
+        assertEquals("28.0", getItemByName("math"), "Your math is incorrect");
```
Comment (Hedgemon4) [inline src/pre_defined_classes/mathematical_program/hard/q4/CalcBillTest.java:24]: This test would never fail, as Clause only accepts 28, so it is redundant if the potential values are not widened

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/pre_defined_classes/mathematical_program/hard/q4/CalcBillTest.java (hunk 1 above)
Comment (SamamthaH) [reply src/pre_defined_classes/mathematical_program/hard/q4/CalcBillTest.java:24]: It's more about providing better feedback than explicitly needing to check the math. If the math is wrong but I didn't have the test they might think their format is the only thing that's wrong.

**Review by SanaBob — APPROVED** [review]

File: src/pre_defined_classes/mathematical_program/hard/q5/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package pre_defined_classes.mathematical_program.hard.q5;
```
[the remark this answers is not in the document]
Comment (SamamthaH) [reply src/pre_defined_classes/mathematical_program/hard/q5/MainTest.java:1]: Iiiiiii have absolutely no idea I'm gonna get rid of it

File: src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package pre_defined_classes.mathematical_program.hard.q4;
+
+/*
+Make a Java program that determines the cost of some take-out sushi, given that you tip at 15% and want to round your
```
Comment (Hedgemon4) [inline src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java:4]: Maybe we should add that the cost of the sushi is 23.75 to the question

Comment (Hedgemon4) [reply src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java:4]: As this variable is just called total, which I think is a bit confusing

File: src/pre_defined_classes/mathematical_program/hard/q4/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        };
+    }
+
+    public void runMain() {
+        CalcBill.main(new String[]{});
+    }
+
+    @Test
+    public void tipPercentageTest() {
+        assertEquals(28.0, Double.parseDouble(getItemByName("billAmount")), 0.1, "Your math is incorrect");
```
Comment (Hedgemon4) [inline src/pre_defined_classes/mathematical_program/hard/q4/MainTest.java:25]: Again, could say something more like "calculated bill amount is incorrect", but I think this is my preference

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
I think the question could be a bit more clear, see my comments

File: src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java (hunk 3 above)
Comment (SamamthaH) [reply src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java]: I have made some modifications to the variable names and question description which should make it look clearer. I'd appreciate it if you could look it over to see if it makes more sense now.

File: src/pre_defined_classes/mathematical_program/hard/q4/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        };
+    }
+
+    public void runMain() {
+        CalcBill.main(new String[0]);
+    }
+
+    @Test
+    public void tipPercentageTest() {
+        assertEquals(28.0, Double.parseDouble(getItemByName("billAmount")), 0.1, "Your final bill calculation is incorrect.");
```
Comment (Hedgemon4) [inline src/pre_defined_classes/mathematical_program/hard/q4/MainTest.java:25]: I think your delta should be smaller here, I don't see it as a problem for this question, but in practice I think that is too wide a range

File: src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java (hunk 3 above)
Comment (Hedgemon4) [reply src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java]: I think this is much better now

Comment (Hedgemon4) [reply src/pre_defined_classes/mathematical_program/hard/q4/CalcBill.java]: Did you update the document with this as well?

**Review by Hedgemon4 — APPROVED** [review]
Looks good

**Review by aadiraju — APPROVED** [review]
Looks good to me!

**Review by Opeyem1a — APPROVED** [review]

**Review by Opeyem1a — APPROVED** [review]

**Pull request merged**